src/org/omegat/filters2/html2/FilterVisitor.java

```diff
@@ -153,38 +162,22 @@ public boolean shouldRecurseChildren() {
     @Override
     public void visitTag(Tag tag) {
 
-        boolean keepIntact = keepIntact(tag);
-
-        if (!keepIntact) {
-            // Decide whether this tag should be kept intact, based on the key-value pairs stored in the
-            // configuration
-            Vector<?> tagAttributes = tag.getAttributesEx();
-            Iterator<?> i = tagAttributes.iterator();
-            while (i.hasNext() && !keepIntact) {
-                Attribute attribute = (Attribute) i.next();
-                String name = attribute.getName();
-                String value = attribute.getValue();
-                if (name == null || value == null) {
-                    continue;
-                }
-                keepIntact = this.filter.checkIgnoreTags(name, value);
-            }
-        }
+        boolean keepIntact = isProtectedTag(tag);
 
         if (keepIntact) {
-            if (text) {
+            if (isTextUpForCollection) {
                 endup();
             } else {
-                flushbefors();
+                writeOutPrecedingNodes();
             }
             writeout(tag.toHtml());
-            if (tag.getEndTag() != null) {
-                recurse = false;
+            if (hasAnEndTag(tag)) {
+                recurseSelf = true;
+                recurseChildren = false;
             }
         } else {
-            // recurse = true;
-            if (isParagraphTag(tag) && text) {
-                endup();
+            if (isParagraphTag(tag)) {
+                handleParagraphTag();
```

Comment:
Here is the difference when compared to the other PR (#402): `handleParagraphTag` handles the check that I removed here (`&& text`) and sets recurseChildren to true.